bookwyrm/templates/snippets/booklist-pagination.html

```diff
@@ -0,0 +1,58 @@
+            <span class="pagination-ellipsis">&hellip;</span>
+            <a class="pagination-link" href="{% url 'list' list.id %}?{% for k, v in request.GET.items %}{% if k != 'page' %}{{ k }}={{ v }}&{% endif %}{% endfor %}page={{ page.previous_page_number }}{{ anchor }}">{{ page.previous_page_number }}</a>
+        {% endif %}
+    <a class="pagination-link is-link" disabled href="{% url 'list' list.id %}?{% for k, v in request.GET.items %}{% if k != 'page' %}{{ k }}={{ v }}&{% endif %}{% endfor %}page={{ page.number }}{{ anchor }}">{{ page.number }}</a>
+        {% if page.number|add:1 == page.paginator.num_pages %}
+            <a class="pagination-link" href="{% url 'list' list.id %}?{% for k, v in request.GET.items %}{% if k != 'page' %}{{ k }}={{ v }}&{% endif %}{% endfor %}page={{ page.next_page_number }}{{ anchor }}">{{ page.next_page_number }}</a>
+        {% elif page.number|add:2 == page.paginator.num_pages %}
+            <a class="pagination-link" href="{% url 'list' list.id %}?{% for k, v in request.GET.items %}{% if k != 'page' %}{{ k }}={{ v }}&{% endif %}{% endfor %}page={{ page.next_page_number }}{{ anchor }}">{{ page.next_page_number }}</a>
+            <a class="pagination-link" href="{% url 'list' list.id %}?{% for k, v in request.GET.items %}{% if k != 'page' %}{{ k }}={{ v }}&{% endif %}{% endfor %}page={{ page.next_page_number|add:1 }}{{ anchor }}">{{ page.next_page_number|add:1 }}</a>
+        {% elif page.number|add:3 == page.paginator.num_pages %}
+            <a class="pagination-link" href="{% url 'list' list.id %}?{% for k, v in request.GET.items %}{% if k != 'page' %}{{ k }}={{ v }}&{% endif %}{% endfor %}page={{ page.next_page_number }}{{ anchor }}">{{ page.next_page_number }}</a>
+            <a class="pagination-link" href="{% url 'list' list.id %}?{% for k, v in request.GET.items %}{% if k != 'page' %}{{ k }}={{ v }}&{% endif %}{% endfor %}page={{ page.next_page_number|add:1 }}{{ anchor }}">{{ page.next_page_number|add:1 }}</a>
+            <a class="pagination-link" href="{% url 'list' list.id %}?{% for k, v in request.GET.items %}{% if k != 'page' %}{{ k }}={{ v }}&{% endif %}{% endfor %}page={{ page.next_page_number|add:2 }}{{ anchor }}">{{ page.next_page_number|add:2 }}</a>
+        {% elif page.number|add:3 < page.paginator.num_pages %}
+            <a class="pagination-link" href="{% url 'list' list.id %}?{% for k, v in request.GET.items %}{% if k != 'page' %}{{ k }}={{ v }}&{% endif %}{% endfor %}page={{ page.next_page_number }}{{ anchor }}">{{ page.next_page_number }}</a>
+            <span class="pagination-ellipsis">&hellip;</span>
+            <a class="pagination-link" href="{% url 'list' list.id %}?{% for k, v in request.GET.items %}{% if k != 'page' %}{{ k }}={{ v }}&{% endif %}{% endfor %}page={{ items.paginator.num_pages }}{{ anchor }}">{{ items.paginator.num_pages }}</a>
```

Comment:
this level of repetition is going to be challenging to maintain. I wonder if it would be possible to get the range that's going to be shown in python and then just loop over that range, checking what should be disabled and whether to include an ellipsis inline?

even just moving the logic into a snippet might make it simpler to read and modify

Reply:
Django 3.2 has this logic as a method on the Paginator: https://docs.djangoproject.com/en/3.2/ref/paginator/#django.core.paginator.Paginator.get_elided_page_range

So I think that in the long term, we'll just end up using that whenever upgrading to that version. So the question is about when that is, and what to do until then. If we don't want to go the route of putting the logic in the template, we could subclass the 3.1 paginator to early bring in the behavior of `get_elided_page_range` from 3.2.

Reply:
I just looked at the release notes for 3.2 and I think upgrading will be trivial. I'm checking now.

Reply:
If this is an imposition, let me know, but I wonder if it would be possible to separate the pagination change into another pull request -- using the existing snippet with just previous/next? It isn't directly related to list sorting, and there are a lot of nuances to implementing it.

Reply:
That plan works for me.

Reply:
That snippet has now been removed.